### PR TITLE
Add logging abstraction

### PR DESCRIPTION
**Pull request by slang25**
This is using `Microsoft.Extensions.Logging.Abstractions`, we could change this for our own interface, however I think the `ILogger`, `ILoggerFactory` & `ILoggerProvider` pattern is really nice, [this post covers it quite well](https://wildermuth.com/2016/04/22/Implementing-an-ASP-NET-Core-RC1-Logging-Provider), and we would just be reinventing the wheel.

This introduces a breaking change to the fluent api, now:
```
CreateMeABus.InRegion(...).With...
```
becomes
```
CreateMeABus.WithLogging(someLoggerFactory).InRegion(...).With...
```

If not logging is required then the consuming code should do something like:
```
var emptyLoggerFactory = new LoggerFactory();
CreateMeABus.WithLogging(emptyLoggerFactory).InRegion(...).With...
```

There are many providers out there, [here is what we'd use for NLog](https://github.com/NLog/NLog.Extensions.Logging)

### CODE REVIEW

File: JustSaying/IHaveFulfilledLoggingRequirements.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying
+{
+    public class JustSayingFleuntlyLogging
```
Comment (AnthonySteele) [inline JustSaying/IHaveFulfilledLoggingRequirements.cs]: typo

File: JustSaying/CreateMeABus.cs [hunk 2]
```diff
@@ -13,33 +12,9 @@ public static class CreateMeABus
         /// </summary>
         public static Func<IAwsClientFactory> DefaultClientFactory = () => new DefaultAwsClientFactory();
 
-        public static IMayWantOptionalSettings InRegion(string region) => InRegions(region);
+        public static JustSayingFleuntlyLogging WithLogging(ILoggerFactory loggerFactory) =>
```
Comment (AnthonySteele) [inline JustSaying/CreateMeABus.cs]: Do a find replace "Fleuntly" to "Fluently"

File: JustSaying.Logging/LoggerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Logging
+{
+    public static class LoggerExtensions
```
Comment (AnthonySteele) [inline JustSaying.Logging/LoggerExtensions.cs]: Longer term we could inline these methods, as it's just a shim for existing methods to ILogger ?

**Comment by AnthonySteele** [comment]
LGTM 👍 What do the expected on .Net core logging think?

File: JustSaying.Logging/LoggerExtensions.cs (hunk 3 above)
Comment (slang25) [inline JustSaying.Logging/LoggerExtensions.cs]: Actually, I think I will inline this, cause I'm throwing out the rest of this project, this was just here to ease the migration. I'll do it as my last commit of this review to reduce the noise.

**Comment by slang25** [comment]
Comments addressed 👍

File: JustSaying.Logging/LoggerExtensions.cs (hunk 3 above)
Comment (AnthonySteele) [inline JustSaying.Logging/LoggerExtensions.cs]: 👍

File: JustSaying/project.json [hunk 4]
```diff
@@ -37,10 +37,10 @@
     "AWSSDK.SimpleNotificationService": "3.3.0.6",
     "AWSSDK.SQS": "3.3.1.6",
     "JustSaying.AwsTools": "5.0.0",
+    "JustSaying.Logging": "5.0.0",
```
Comment (martincostello) [inline JustSaying/project.json]: Looks like the AppVeyor build fails trying to restore this reference.

File: JustSaying/JustSayingFluentlyExtensions.cs [hunk 5]
```diff
@@ -1,11 +1,44 @@
...
+using JustSaying.Messaging.Monitoring;
 using JustSaying.Models;
 
 namespace JustSaying
 {
     public static class JustSayingFluentlyExtensions
     {
+        public static IMayWantOptionalSettings InRegion(this JustSayingFluentlyLogging logging, string region) => InRegions(logging, region);
+
+        public static IMayWantOptionalSettings InRegions(this JustSayingFluentlyLogging logging, params string[] regions)
```
Comment (martincostello) [inline JustSaying/JustSayingFluentlyExtensions.cs]: Is it worth adding an overload that also takes ```IEnumerable<string>``` for the regions?

File: JustSaying/project.json (hunk 4 above)
Comment (martincostello) [inline JustSaying/project.json]: Looks like this is broken in master too actually...

Comment (slang25) [inline JustSaying/project.json]: Yeah, this is broken, but you are right the build is broken atm anyway. Good spot

File: JustSaying/JustSayingFluentlyExtensions.cs (hunk 5 above)
Comment (slang25) [inline JustSaying/JustSayingFluentlyExtensions.cs]: We could do, I'm not sure it adds much though, I presume it's done like this to enable `...InRegions("uk","fr")` as well as `...InRegions(tenants.ToArray())`.
Yeah, let me do that unless anyone disagrees.

File: JustSaying/JustSayingFluentlyExtensions.cs [hunk 6]
```diff
@@ -12,6 +14,8 @@ public static class JustSayingFluentlyExtensions
     {
         public static IMayWantOptionalSettings InRegion(this JustSayingFluentlyLogging logging, string region) => InRegions(logging, region);
 
+        public static IMayWantOptionalSettings InRegions(this JustSayingFluentlyLogging logging, IEnumerable<string> regions) => InRegions(logging, regions.ToArray());
```
Comment (martincostello) [inline JustSaying/JustSayingFluentlyExtensions.cs]: Does it work OK with overload resolution if the implementation is in ```IEnumerable<string>``` and the ```string[]``` overload calls it? Then we don't need the ```ToArray()``` call.

Comment (martincostello) [inline JustSaying/JustSayingFluentlyExtensions.cs]: I suppose that would just create a stack overflow without a cast, but a cast would be more lightweight than a copy?

**Comment by martincostello** [comment]
I just noticed there's a section about logging in the README. Should we update it in this PR or do that separately?

**Comment by martincostello** [comment]
The CI looks like it's hanging now having issues loading NUnit:

[```Could not load file or assembly 'nunit.framework' or one of its dependencies. The system cannot find the file specified.```](https://ci.appveyor.com/project/justeattech/justsaying/build/4.0.226#L493)

File: JustSaying/JustSayingFluentlyExtensions.cs (hunk 6 above)
Comment (slang25) [inline JustSaying/JustSayingFluentlyExtensions.cs]: Yeah, so arrays will pick the `string[]` overload, things like `List<string>` will use the other overload, all works as expected. Unless you mean reversing the overloads, which makes more sense, we'd need to use `.AsEnumerable()` to not create an infinite recursive call.

Comment (martincostello) [inline JustSaying/JustSayingFluentlyExtensions.cs]: Yeah I meant reversing and using a cast/```as``` so a copy isn't needed:

```csharp
public static IMayWantOptionalSettings InRegion(this JustSayingFluentlyLogging logging, params string[] regions) => InRegions(logging, regions as IEnumerable<string>);

public static IMayWantOptionalSettings InRegions(this JustSayingFluentlyLogging logging, IEnumerable<string> regions)
{
    // Stuff
}
```

**Comment by martincostello** [comment]
Something's still making it hang :/

**Comment by slang25** [comment]
Yeah, it's AppVeyor selecting NUnit 3.5 automatically, we have to change it to use a script. If it works, I'll revert it and add an `appveyor.yml` to this branch.

**Comment by slang25** [comment]
Wait, I'm being dumb. It is using an `appveyor.yml` but it might be using the one from master.

**Comment by martincostello** [comment]
I thought it always used the one from the branch it builds.

File: appveyor.yml [hunk 7]
```diff
@@ -1,41 +1,44 @@
...
-  on_build_failure: true
-  on_build_status_changed: true
-deploy:
-- provider: NuGet
-  api_key:
-    secure: 6MzbzEs4YdJKS67Gio5gEO8mNKmwfC4UHTCmECZ1KOutI6ndm4vAECazmVNB6an7
-  artifact: /.*nupkg/
-  on:
-    APPVEYOR_REPO_TAG: true
+version: 4.0.{build}
```
Comment (AnthonySteele) [inline appveyor.yml]: what's happening with the `^M` chars here?

Comment (slang25) [inline appveyor.yml]: Not sure, that wasn't intentional. I'll try to fix it

**Comment by martincostello** [comment]
Looks good now apart from the ```^M```'s.

**Comment by slang25** [comment]
Thanks. I'm just updating the README file to reflect the changes.

File: JustSaying.IntegrationTests/WhenRegisteringHandlersViaResolver/WhenRegisteringAHandlerViaContainerWithMissingRegistration.cs [hunk 8]
```diff
@@ -1,34 +1,35 @@
...
+        protected override void Given()
+        {
+            RecordAnyExceptionsThrown();
+        }
+
+        protected override Task When()
+        {
+            var handlerResolver = new StructureMapHandlerResolver(new Container());
+
+            CreateMeABus.WithLogging(new LoggerFactory()).InRegion("eu-west-1")
```
Comment (AnthonySteele) [inline JustSaying.IntegrationTests/WhenRegisteringHandlersViaResolver/WhenRegisteringAHandlerViaContainerWithMissingRegistration.cs]: Inconsistent formatting, use one "WithXXX" per line

**Comment by AnthonySteele** [comment]
merge it!

**Pull request merged**